File: src/jev_factorio/successors.py
"""Strict additive-producer evidence. A paid furnace is not a useful factory."""
from __future__ import annotations

import math

ROLES = {'growth:iron-plate': 'iron-ore', 'growth:copper-plate': 'copper-ore'}
PHASES = {'reserved', 'output_building', 'output_commissioning', 'input_building', 'producing', 'preferred'}
USE_RECIPES = {'iron-gear-wheel', 'copper-cable', 'electronic-circuit',
               'automation-science-pack', 'logistic-science-pack'}
MARKER = 'successor_project'
MAX_PROJECT_TICKS = 216000
TRIAL_LIMIT = 200
SCIENCE_RESERVE = 20


def integer(value, minimum=0, maximum=2**53-1):
    return type(value) is int and minimum <= value <= maximum


def text(value):
    return isinstance(value, str) and 0 < len(value) <= 128


def sources(snapshot) -> dict:
    data = snapshot.factory.get('successors')
    if (not isinstance(data, dict) or not integer(data.get('protocol'), 1, 1)
            or data.get('session_id') != snapshot.session_id or not integer(data.get('tick'))
            or data['tick'] != snapshot.tick or not isinstance(data.get('sources'), dict)
            or set(data['sources']) - ROLES.keys()):
        raise ValueError('Missing or stale successor evidence')
    entities = snapshot.factory.get('entities', {})
    for role, row in data['sources'].items():
        item = role[7:]
        if (not isinstance(row, dict) or row.get('source') != role or row.get('item') != item
                or row.get('predecessor') != 'recipe:' + item or not text(row.get('anchor'))
                or not row['anchor'].startswith('cell-site:') or row.get('phase') not in PHASES
                or not integer(row.get('predecessor_unit'), 1) or not integer(row.get('source_unit'))
                or not integer(row.get('paid'), 0, 1) or not integer(row.get('seeded'), 0, 10)
                or not integer(row.get('trial_collected')) or not integer(row.get('credit'))
                or not integer(row.get('attribution_resets')) or not integer(row.get('started_tick'))
                or row['started_tick'] > snapshot.tick or not integer(row.get('remaining_ore'))
                or not isinstance(row.get('use'), dict) or not isinstance(row.get('qualification'), dict)
                or row.get('fault')):
            raise ValueError('Invalid or faulted successor evidence')
        old = entities.get(row['predecessor'], {})
        if old.get('unit_number') != row['predecessor_unit'] or old.get('name') != 'stone-furnace':
            raise ValueError('Successor predecessor identity changed')
        if row['source_unit']:
            entity = entities.get(role, {})
            site = snapshot.factory.get('production_sites', {}).get('sources', {}).get(role, {})
            if (row['source_unit'] == row['predecessor_unit'] or row['paid'] != 1
                    or entity.get('unit_number') != row['source_unit'] or entity.get('name') != 'stone-furnace'
                    or site.get('source_unit') != row['source_unit'] or site.get('anchor') != row['anchor']
                    or site.get('state') != 'owned' or entity.get('position') != site.get('position')):
                raise ValueError('Successor owned producer changed')
        elif row['paid'] or row['phase'] != 'reserved':
            raise ValueError('Unpaid successor cannot progress')
        use = row['use']
        if use and (not text(use.get('job_id')) or use.get('recipe') not in USE_RECIPES
                    or not integer(use.get('quantity'), 3, TRIAL_LIMIT * 200)
                    or use.get('source_unit') != row['source_unit'] or type(use.get('source_unit')) is not int
                    or not text(use.get('input_layout')) or not integer(use.get('completed_tick'))
                    or not row['started_tick'] <= use['completed_tick'] <= snapshot.tick
                    or not integer(use.get('requested'), 1, 200) or type(use.get('finished')) is not int
                    or use['finished'] != use['requested'] or not isinstance(use.get('outputs'), dict)
                    or not use['outputs'] or any(not text(k) or not integer(v, 1) for k, v in use['outputs'].items())):
            raise ValueError('Invalid downstream craft witness')
        proof = row['qualification']
        if proof and (not use or any(not integer(proof.get(key)) for key in
                ('first_tick', 'last_tick', 'positive_samples', 'produced', 'source_unit', 'last_progress_tick', 'max_observation_gap'))
                or proof['source_unit'] != row['source_unit'] or proof.get('input_layout') != use['input_layout']
                or proof.get('use_job_id') != use['job_id'] or proof['positive_samples'] < 3
                or proof['produced'] < 3 or proof['last_tick'] - proof['first_tick'] < 36000
                or not proof['first_tick'] <= proof['last_progress_tick'] <= proof['last_tick']
                or proof['last_tick'] - proof['last_progress_tick'] > 1800 or proof['max_observation_gap'] > 1800
                or not row['started_tick'] <= proof['first_tick'] <= proof['last_tick'] <= snapshot.tick):
            raise ValueError('Invalid successor qualification')
        if row['phase'] == 'preferred' and not proof:
            raise ValueError('Preferred successor lacks qualification')
    return data['sources']


def flowing(role, snapshot) -> bool:
    from .input_routes import flow_complete
    from .output_buffers import flow_complete as output_complete
    route = snapshot.factory.get('input_routes', {}).get('sources', {}).get(role, {})
    output = snapshot.factory.get('output_buffers', {}).get('sources', {}).get(role, {})
    return flow_complete(role, route.get('layout', ''), snapshot) and output_complete(role, output.get('layout', ''), snapshot)


def qualified(role, snapshot) -> bool:
    row = sources(snapshot).get(role)
    if not row or row['phase'] != 'preferred' or not row['qualification'] or not flowing(role, snapshot):
        return False
    route = snapshot.factory['input_routes']['sources'][role]
    return row['qualification']['input_layout'] == route['layout'] and row['use']['input_layout'] == route['layout']


def private_output(role, snapshot) -> bool:
    """Only the successor-aware planner chooses trial/preferred collections."""
    if role in ROLES:
        return True
    for source, row in snapshot.factory.get('output_buffers', {}).get('sources', {}).items():
        if source in ROLES and role == row.get('chest_role'):
            return True
    return False


def permits(action, parameters, snapshot) -> bool:
    rows = sources(snapshot)
    role = parameters.get('role')
    if action == 'factory_place' and role in ROLES:
        site = snapshot.factory.get('production_sites', {}).get('sources', {}).get(role, {})
        old = snapshot.factory.get('entities', {}).get('recipe:' + role[7:], {})
        return bool(parameters.get('anchor') == site.get('anchor') and site.get('state') == 'proposed'
                    and old.get('name') == 'stone-furnace' and old.get('products_finished', 0) >= 20
                    and all(snapshot.inventory.get(k, 0) >= v for k, v in initial_kit(site, ROLES[role]).items()))
    for source, row in rows.items():
        if role == source:
            if action in {'factory_extract', 'factory_configure'}:
                return False
            if action == 'factory_insert' and parameters.get('item') != 'coal':
                return (parameters.get('item') == ROLES[source]
                        and row['seeded'] + parameters.get('quantity', TRIAL_LIMIT) <= 10
                        and source not in snapshot.factory.get('input_routes', {}).get('sources', {}))
        output = snapshot.factory.get('output_buffers', {}).get('sources', {}).get(source, {})
        if role == output.get('chest_role') and action == 'factory_extract':
            return bool(flowing(source, snapshot) and (qualified(source, snapshot)
                        or row['trial_collected'] + parameters['quantity'] <= TRIAL_LIMIT))
    return True


def initial_kit(site, ore) -> dict:
    bill = dict(site.get('bill', {}))
    if not bill or ore not in ROLES.values():
        raise ValueError('Missing successor bill')
    bill['transport-belt'] = bill.get('transport-belt', 0) + SCIENCE_RESERVE
    bill.update(coal=30)
    bill[ore] = 10
    return bill


def project_valid(project, role, tick):
    if (role not in ROLES or not isinstance(project, dict)
            or set(project) != {'anchor', 'predecessor_unit', 'source_unit', 'started_tick', 'deadline_tick', 'status'}
            or not text(project['anchor']) or not project['anchor'].startswith('cell-site:')
            or not integer(project['predecessor_unit'], 1) or not integer(project['source_unit'])
            or not integer(project['started_tick']) or not integer(project['deadline_tick'])
            or not project['started_tick'] <= tick
            or project['deadline_tick'] != project['started_tick'] + MAX_PROJECT_TICKS
            or project['status'] not in {'active', 'qualified', 'paused'}):
        raise ValueError('Invalid retained successor project')

File: src/jev_factorio/lua/successors.lua
-- Additive owned producers. No resource, entity, speed or control writes here.
-- Existing fair builders/transfers/craft jobs remain the only mutation authority.
local c,fair=assert(storage.campaign),assert(storage.fair)
assert(storage.input_routes and storage.output_buffers and storage.production_sites,
    "Successors require owned input routes and production sites")
assert(c.craft_jobs and c.begin_craft_job,"Successors require receipt-tracked crafting")
assert(not storage.mining_outposts,"Successor/outpost composition is not yet supported")
local s=storage.successors or {protocol=1,records={},pending_use=nil}
assert(s.protocol==1,"Unsupported successor runtime")
storage.successors=s;c.successors_enabled=true
local roles={["growth:iron-plate"]="iron-ore",["growth:copper-plate"]="copper-ore"}
local useful={['iron-gear-wheel']=true,['copper-cable']=true,['electronic-circuit']=true,
    ['automation-science-pack']=true,['logistic-science-pack']=true}
local function integer(n) return type(n)=="number" and n%1==0 and n>=0 and n<9007199254740992 end
local function predecessor(role) return "recipe:"..string.sub(role,8) end
local function identity(role,m)
    local p=fair.actor();local old=c.entities[predecessor(role)]
    assert(old and old.valid and old.unit_number==m.predecessor_unit and old.name=="stone-furnace"
        and old.surface==p.surface and old.force==p.force,"Successor predecessor identity changed")
    if m.source_unit>0 then
        local e=c.entities[role];local site=storage.production_sites.owned[role]
        assert(e and e.valid and e.unit_number==m.source_unit and e~=old and site and site.entity==e
            and site.anchor==m.anchor and e.force==p.force and e.surface==p.surface,
            "Successor identity changed")
    end
    return p
end
c.successor_admit=function(site,building)
    if not roles[site.role] then return end
    local p=fair.actor();local old=c.entities[predecessor(site.role)]
    assert(old and old.valid and old.name=="stone-furnace" and old.products_finished>=20,
        "An established predecessor is required")
    local m=s.records[site.role]
    if m then assert(not m.fault and m.anchor==site.anchor,"Successor intent changed");identity(site.role,m) end
    if building then assert(m and m.source_unit==0,"Successor furnace must be prepared once") end
    for role,other in pairs(s.records) do
        assert(role==site.role or other.qualification~=nil,"Only one unqualified successor project is allowed")
    end
    local kit={["stone-furnace"]=1,["burner-mining-drill"]=1,["burner-inserter"]=2,
        ["wooden-chest"]=1,["transport-belt"]=site.belt_count+20,coal=30,[site.ore]=10}
    for name,count in pairs(kit) do assert(p.get_item_count(name)>=count,"Successor full paid kit or science reserve missing") end
end
c.successor_prepare=function(site)
    if not roles[site.role] then return end
    c.successor_admit(site,false)
    if not s.records[site.role] then
        s.records[site.role]={anchor=site.anchor,predecessor_unit=c.entities[predecessor(site.role)].unit_number,
            source_unit=0,paid=0,seeded=0,seed_uncollected=0,trial_collected=0,credit=0,
            expected_count=fair.actor().get_item_count(site.item),attribution_resets=0,
            started_tick=game.tick}
    end
end
c.successor_bound=function(site)
    if not roles[site.role] then return end
    local m=assert(s.records[site.role],"Missing prepared successor")
    assert(m.source_unit==0 and site.source_unit~=m.predecessor_unit,"Duplicate successor bind")
    m.source_unit=site.source_unit;m.paid=1;identity(site.role,m)
end
local function input_ready(role)
    local route=storage.input_routes.cells[role]
    return route and not route.fault and route.flow and route.flow.conservation==true
end
local function synchronize(m,before)
    if m.expected_count~=before then m.credit=0;m.attribution_resets=m.attribution_resets+1 end
end
c.successor_before_transfer=function(role,item,quantity,extracting)
    local p=fair.actor()
    for source,m in pairs(s.records) do
        identity(source,m);assert(not m.fault,"Successor requires reconciliation")
        local out=storage.output_buffers.cells[source]
        if role==source then
            assert(not extracting,"Do not manually extract successor furnace output")
            if item~="coal" then
                assert(item==roles[source] and not input_ready(source) and m.seeded+quantity<=10
                    and not storage.input_routes.cells[source],"Only bounded output-commissioning seed is permitted")
            end
        end
        if out and role==out.chest_role then
            assert(extracting and input_ready(source),"Successor output requires ore-to-plate flow")
            assert(m.qualification or m.trial_collected+quantity<=200,"Successor trial collection budget exhausted")
        end
    end
    return p.get_item_count(item)
end
c.successor_after_transfer=function(role,item,quantity,receipt,extracting,before)
    local p=fair.actor();local after=p.get_item_count(item)
    assert(after-before==(extracting and quantity or -quantity),"Successor attribution transfer mismatch")
    for source,m in pairs(s.records) do
        if role==source and not extracting and item==roles[source] then
            m.seeded=m.seeded+quantity;m.seed_uncollected=m.seed_uncollected+quantity
        end
        if item==string.sub(source,8) then
            synchronize(m,before)
            local out=storage.output_buffers.cells[source]
            if extracting and out and role==out.chest_role and input_ready(source) then
                local old=math.min(quantity,m.seed_uncollected)
                m.seed_uncollected=m.seed_uncollected-old
                m.credit=m.credit+quantity-old;m.trial_collected=m.trial_collected+quantity
            elseif not extracting then
                local attributable=math.max(0,quantity-(before-m.credit))
                m.credit=math.max(0,m.credit-attributable)
            end
            m.expected_count=after
        end
    end
end
c.successor_craft_paid=function(job,before)
    local p=fair.actor();local claims={}
    for source,m in pairs(s.records) do
        local item=string.sub(source,8);local count=job.inputs[item]
        if count then
            synchronize(m,before[item])
            local attributable=math.max(0,count-(before[item]-m.credit))
            m.credit=math.max(0,m.credit-attributable);m.expected_count=p.get_item_count(item)
            if attributable>0 and useful[job.recipe] and input_ready(source) then
                claims[source]={quantity=attributable,source_unit=m.source_unit,
                    layout=storage.input_routes.cells[source].layout}
            end
        end
    end
    -- Losing attribution is conservative: never reinterpret a different job as
    -- completion of an older one, and never change its native job status.
    if next(claims) then s.pending_use={job_id=job.id,claims=claims} else s.pending_use=nil end
end
c.observe_successors=function(result)
    local p=fair.actor();local pending=s.pending_use;local job=result.craft_job
    if pending and job and job.id==pending.job_id and job.status=="completed" and job.paid
        and job.queue_valid and job.accepted==job.requested and job.finished==job.requested
        and job.session_id==storage.jev_session_id and job.player_index==p.index
        and job.unit_number==p.character.unit_number and job.surface_index==p.surface.index
        and job.force_index==p.force.index then
        local complete=true
        for item,count in pairs(job.outputs) do
            if p.get_item_count(item)<job.baseline[item]+count then complete=false end
        end
        if complete then
            for source,claim in pairs(pending.claims) do
                local m=s.records[source];local route=storage.input_routes.cells[source]
                if m and route and input_ready(source) and m.source_unit==claim.source_unit
                    and route.layout==claim.layout and claim.quantity>=3 and not m.use then
                    m.use={job_id=job.id,recipe=job.recipe,quantity=claim.quantity,source_unit=m.source_unit,
                        input_layout=route.layout,completed_tick=job.completed_tick,
                        requested=job.requested,finished=job.finished,outputs=job.outputs}
                end
            end
            s.pending_use=nil
        end
    end
    local rows={}
    for role,m in pairs(s.records) do
        local ok=pcall(identity,role,m)
        if not ok then m.fault="successor_identity_changed" end
        local source=c.entities[role];local input=result.input_routes.sources[role]
        local output=result.output_buffers.sources[role]
        local phase="reserved";local remaining=0
        if m.source_unit>0 then phase="output_building" end
        if output and next(output.parts) and not next(output.flow) then phase="output_commissioning" end
        if output and next(output.flow) then phase="input_building" end
        local flowing=input and input.state=="ready" and input.topology and next(input.flow)
        if flowing then
            phase="producing"
            local native=storage.input_routes.cells[role]
            if native and native.patch then for _,ore in ipairs(native.patch) do if ore.valid then remaining=remaining+ore.amount end end end
            local count=source.products_finished
            local window=m.window
            if not integer(count) or window and count<window.last_produced then m.fault="successor_counter_regressed"
            elseif not m.qualification then
                if not window or game.tick-window.last_tick>1800 or game.tick-window.last_progress_tick>1800 then
                    m.window={first_tick=game.tick,last_tick=game.tick,baseline=count,last_produced=count,positive_samples=0,last_progress_tick=game.tick,max_gap=0}
                    window=m.window
                elseif game.tick-window.last_tick>=60 then
                    window.max_gap=math.max(window.max_gap,game.tick-window.last_tick)
                    if count>window.last_produced then window.positive_samples=window.positive_samples+1;window.last_progress_tick=game.tick end
                    window.last_tick=game.tick;window.last_produced=count
                end
                if window.positive_samples>=3 and window.last_tick-window.first_tick>=36000
                    and count-window.baseline>=3 and m.use then
                    m.qualification={first_tick=window.first_tick,last_tick=window.last_tick,
                        positive_samples=window.positive_samples,produced=count-window.baseline,
                        last_progress_tick=window.last_progress_tick,max_observation_gap=window.max_gap,
                        source_unit=m.source_unit,input_layout=input.layout,use_job_id=m.use.job_id}
                end
            end
            if m.qualification and not m.fault then phase="preferred" end
        end
        if m.fault then phase="fault" end
        local item=string.sub(role,8);synchronize(m,p.get_item_count(item));m.expected_count=p.get_item_count(item)
        rows[role]={source=role,item=item,anchor=m.anchor,predecessor=predecessor(role),
            predecessor_unit=m.predecessor_unit,source_unit=m.source_unit,paid=m.paid,
            phase=phase,seeded=m.seeded,trial_collected=m.trial_collected,credit=m.credit,
            attribution_resets=m.attribution_resets,started_tick=m.started_tick,remaining_ore=remaining,
            use=m.use or {},qualification=m.qualification or {},fault=m.fault}
    end
    return {protocol=1,session_id=storage.jev_session_id,tick=game.tick,sources=rows}
end
